Test model, test taxonomies and unit tests added

innovad opened the pull request:
Please note this is not a complete set of unit tests yet, as it is work in progress.

This pull request includes the closed request https://github.com/liquid-mind/inflection/pull/10

innovad committed:
add some blackbox tests

innovad committed:
AllTests Suite added

innovad committed:
add travis support

innovad committed:
travis: java 8

innovad committed:
build: add jacoco support

innovad committed:
tests: package structure refactored

innovad committed:
tests: rename

innovad committed:
tests: proxy generation added

innovad committed:
tests: proxy and compiler test added

innovad committed:
tests: compiler tests improved

innovad committed:
typo fixed

innovad committed:
tests: blackbox bean tests added

innovad committed:
organize imports

innovad committed:
tests: more blackbox bean tests added

innovad committed:
tests: size() test added

innovad committed:
tests: more compiler tests added

innovad committed:
tests: added test for calculated field

innovad committed:
Merge branch 'inflection-2.0' into feature/blackboxtests

Conflicts:
	build.gradle

innovad committed:
InflectionCompilerTest refactored, naming improved

innovad committed:
TaxonomyTest refactored, naming improved

innovad committed:
TaxonomyTest @Ignore removed

innovad committed:
TaxonomyTest naming improved

innovad committed:
ViewTest refactored, naming improved

innovad committed:
MemberTest refactored, naming improved

innovad committed:
BlackboxTest refactored, naming improved

innovad committed:
BlackboxTest naming improved

innovad committed:
package naming refactored

innovad committed:
TestSuite removed

innovad committed:
order fixed

johnbrush commented on build.gradle:
```diff
@@ -74,7 +75,13 @@ task( compileTestTaxonomiesInflection, dependsOn: compileTestModelJava, type: Ja
 	classpath = sourceSets.main.runtimeClasspath + sourceSets.testModel.runtimeClasspath
 	jvmArgs '-ea'
-	args "${buildDir}/resources/testModel", "BOOTSTRAP", "src/test/resources/ch/liquidmind/inflection/test/model/TestModel.inflect"
+	args "${buildDir}/resources/testModel", "BOOTSTRAP", 
```
I think we should consider running the compiler(s) and proxy generator from within the tests themselves. The workflow in Gradle makes sense for applications that _use_ Inflection, but for our own tests I think it might be better for any errors to show up in the junit results. To make this work we would need to use a custom class loader; we'll also have to compile the Java classes using the Java tools API (analogous to Deflector). BTW: this also makes debugging easier.

innovad replied:
We could do it in @BeforeClass or within the test method itself, depending on the situation. Performance could be an issue. I suggest to think about this and implement it in a subsequent branch.

innovad replied:
I have now updated this pull request with "self contained" tests. Taxonomy compilation, proxy generation, compilation and loading (if required by the test) is now done inside the test class, generally in a @BeforeClass method for each test class. Commit https://github.com/innovad/inflection/commit/fe57f3b2de1f41fcd64401258486997f56ede81f

johnbrush commented on src/main/java/ch/liquidmind/inflection/proxy/ProxyGenerator.java:
```diff
@@ -40,6 +40,10 @@ public static void main( String[] args )
 		for ( String taxonomyName : taxonomyNames )
 		{
 			Taxonomy taxonomy = TaxonomyLoader.getContextTaxonomyLoader().loadTaxonomy( taxonomyName );
+			if (taxonomy == null) {
```
I'm not sure that this is the way I would handle it; I may introduce a general purpose ProxyGeneratorException (and InflectionCompilerException), but I need to go through the code and have a good think. I appreciate your initiative, but could you take it out for now? You're welcome to define a test that fails so that don't forget about it.

innovad replied:
I have replaced the change in ProxyGenerator with a TODO, so we can find a solution later

innovad committed:
replaced change in ProxyGenerator with TODO

innovad committed:
added text to Ignore annotation

johnbrush commented on src/test/java/ch/liquidmind/inflection/compiler/InflectionCompilerTest.java:
```diff
@@ -0,0 +1,55 @@
+import ch.liquidmind.inflection.compiler.CompilationJob.CompilationMode;
+import ch.liquidmind.inflection.compiler.util.CompilerTestUtility;
+
+public class InflectionCompilerTest {
```
Need to format the code using "standard" ZHAW IWI beautifier; I'm sending you the eclipse files now. Also: could you see about integrating this with gradle? I know that there is a plugin that reads eclipse formatter files---Alpnet used it in ExpAct, but it wasn't working, so I took it out. I assume it does, however, work if configured correctly.

johnbrush commented on src/test/java/ch/liquidmind/inflection/compiler/InflectionCompilerTest.java:
```diff
@@ -0,0 +1,55 @@
+
+	@Test
+	public void testCompile_ValidFile_SuccessfulCompilation() throws Exception {
+		CompilationJob job = CompilerTestUtility.createCompilationJob(CompilationMode.BOOTSTRAP, 
```
Just a question; do you understand what the CompilationMode.BOOTSTRAP Option does? It's not a trick question; I just wanted to make sure you know what the difference is.

innovad replied:
I did see that compilation mode has an influence on reserved names (e.g. "ch.liquidmind.inflection") in Pass1Listener#validateDoesntUseReservedNames()

Are there any other Ideas behind it? Perhaps a small JavaDoc on CompilationMode values would be great :smile:

johnbrush replied:
Yes, yes, I know… all in good time ;-) The reserved names is the key point (as far as I remember).

innovad committed:
ZHAW Cleanup/Formatter applied.

johnbrush commented on src/test/java/ch/liquidmind/inflection/compiler/InflectionCompilerTest.java:
```diff
@@ -0,0 +1,55 @@
+	@Test
+	public void testCompile_ValidFile_SuccessfulCompilation() throws Exception {
+		CompilationJob job = CompilerTestUtility.createCompilationJob(CompilationMode.BOOTSTRAP, 
+				this.getClass().getResource("InflectionCompilerTest_ValidFile.inflect"));
```
I would introduce a new utility method "getClassLoader" that returns the class loader to use for testing. I would also recommend using Thread.getContextClassLoader().getResource() instead of Class.getResource; the rationale is that the class loader of the test class might not necessarily be the class loader through which the resource can be located. If we start running the compiler and proxy generator through custom class loaders, this may be relevant.

innovad committed:
CompilerTestUtility refactored, simplified usage

innovad committed:
Merge branch 'inflection-2.0' into feature/blackboxtests

innovad committed:
Merge remote-tracking branch 'origin/inflection-2.0' into
feature/blackboxtests

Conflicts:
	build.gradle

innovad committed:
self-contained tests

tests are now fully self-contained:
	- compilation, proxy-generation and loading is done inside test class

innovad committed:
Merge remote-tracking branch 'origin/inflection-2.0' into
feature/blackboxtests

Conflicts:
	build.gradle

innovad committed:
tabs to spaces

innovad committed:
fix travis-ci gradlew problem
